DAOS-7203 control: Add histogram support to Prometheus exporter

mjmac opened the pull request:
Update the Prometheus exporter to support passthrough histograms
from native DAOS telemetry format. Fixes a few bugs and inefficiencies
in the native histogram implementation.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mjmac commented:
<img width="1259" alt="Screenshot 2024-02-16 at 20 33 45" src="https://github.com/daos-stack/daos/assets/83737/867b8ea3-6f1f-44c6-a916-b0515bdb5ae0">[
](https://lh6.googleusercontent.com/proxy/db9Q_Pldzlz2MIcPow606X8f10qf_0WFrSqA6q0CD2c82jXY4QpO2jMAIZmGlxEDxKXn4GCJt9_eD-ZIaey8Qr6f2gqDpD95ll4XwvhWnagwE-Tu8w)

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5382/11/execution/node/1177/log

daltonbohning commented:
FYI - I think the dkey_akey_enum_punch.py failures are being fixed in #14301

daltonbohning commented:
> FYI - I think the dkey_akey_enum_punch.py failures are being fixed in #14301

Actually, sorry. That looks like a different issue

mjmac commented:
Note to reviewers... This is finally in a state where I think it's ready for landing. I reworked the patch a bit. The original version changed the fetch/update bytes counters to histograms in addition to adding all of the logic for exporting via Prometheus. I've reverted that part of it, which should make this patch less risky. Longer term, I would like to revisit that other change, but I'll do that separately so that we can evaluate it in isolation.

kjacque reviewed (approved):
Nice work. I just had a couple minor comments, which I wouldn't block on.

tanabarr commented on src/gurt/telemetry.c:
```diff
@@ -3200,6 +3226,8 @@ d_tm_init_histogram(struct d_tm_node_t *node, char *path, int num_buckets,
 		goto failure;
 	}
+	histogram->dth_fast_bucketing =
+	    (multiplier == 2 && (initial_width & (initial_width - 1)) == 0);
```
to me it reads "set dth_fast_bucketing iff mult is 2 and least-significant-bit of initial_width is set, is that correct?

mjmac replied:
I'll add a comment in a subsequent patch. The idea is easier to see with an example. Say `initial_width` is 256. As a bit pattern, that value looks like this: `0b100000000`. Then, if we subtract 1 from it, we get this pattern: `0b11111111`. If we AND those together, we get 0. This only works for numbers that are clean powers of 2. 

Even easier to see if we "stack" the two values:
```
100000000
 11111111
```

kjacque replied:
When you do add the comment, it would be great to detail not only the meaning of the bitwise math (although maybe that would be most clarified by using a named helper function or macro), but also why fast bucketing only works in the case where everything is a power of two. I think detailing the principle behind it will be most helpful in the future.

tanabarr reviewed (approved):
Nice work! nothing blocking

mjmac commented on src/control/lib/telemetry/duration.go:
```diff
@@ -25,8 +25,11 @@ import (
 )
 
+var _ StatsMetric = (*Duration)(nil)
+
```
It's not an exported type, it's an assertion that `Duration` implements `StatsMetric`.

tanabarr replied:
the comment is referring to "Duration" type below
